### PR TITLE
Enable tab-completion for nuget package versions

### PR DESCRIPTION
**Pull request by baronfel**
The tab completion takes into account any fragments of the version entered, as well as the existing --prerelease flag to filter the returned versions.

Demo:
* tab completion of package id based on fragment
* --version completion of full id
* --version completion using --prerelease to allow preview versions
* --version completion using version fragments to narrow versions

name-completion

Cases to cover:
* [x] slow feeds - need some kind of max timeout to prevent hangs
  * chose a 500ms max for the autocomplete requests
* [x] feeds that require auth - filter them out? `--interactive` likely will not work in a completions context
  * handled this by having any feed communication exceptions just return empty completions for that feed, this isn't critical enough to need diagnostics/loggin g

### CODE REVIEW

**Comment by aortiz-msft** [comment]
@martinrrm - Would you please take a look?

**Comment by baronfel** [comment]
@martinrrm if you want to take it for a spin yourself:

* clone this PR
* build with `./build.cmd` or equivalent for your platform
* run the 'dogfood' script: `./eng/dogfood.ps1` or equivalent for your platform
* your `dotnet` is now the freshly-built one from this PR - make sure your shell has [dotnet tab completion](https://learn.microsoft.com/en-us/dotnet/core/tools/enable-tab-autocomplete) enabled and you should be good to go!

**Comment by anangaur** [comment]
This is great addition 👏🏽👏🏽👏🏽

**Comment by JonDouglas** [comment]
image

🚢 IT!!!!!!!

File: src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs [hunk 1]
```diff
@@ -722,6 +722,47 @@ private string GenerateVersionRangeErrorDescription(string packageIdentifier, Ve
             return packageMetadata.Identity.Version;
         }
 
+        public async Task<IEnumerable<NuGetVersion>> GetPackageVersionsAsync(PackageId packageId, string versionPrefix = null, bool allowPrerelease = false, PackageSourceLocation packageSourceLocation = null, CancellationToken cancellationToken = default)
+        {
+            // grab allowed sources for the package in question
+            IEnumerable<PackageSource> packagesSources = LoadNuGetSources(packageId, packageSourceLocation);
```
[the remark this answers is not in the document]
Comment (baronfel) [reply src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:728]: Good spot @martinrrm - that's querying for the package IDs and wasn't changed in this PR but it should be updated. It should be doing something similar to what I've done in this file - loading sources from NuGet config files and finding the search endpoints from those files.

[the remark this answers is not in the document]
Comment (baronfel) [reply src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:743]: @martinrrm I've just pushed an update that also completes package Ids using the same Autocomplete Resource. Everything's nicely aligned now, thanks again!

**Review by martinrrm — APPROVED** [review]
Not sure if you need an approval from me, change looks great, there is just that improvement we can do, not sure if it's going to be done in this PR. Thanks for the change!

**Comment by baronfel** [comment]
> Not sure if you need an approval from me, change looks great, there is just that improvement we can do, not sure if it's going to be done in this PR. Thanks for the change!

I actually did that improvement today! So all is looking good. I'm going to ask the SDK team for code review as well then we can merge it.

**Comment by baronfel** [comment]
> Does this deserve additions to telemetry notices? I don't remember any other tab completion in dotnet contacting network servers.

I don't think it requires any additional telemetry notices - while this is novel in the sense that it checks your configured sources for packages, you're _already_ signalling consent to communicating with those registries during restore, etc. Are there specific kinds of data you're concerned about leaking to feed owners that we can help proactively prevent?

**Comment by baronfel** [comment]
Package Id completions seem to be broken, needs more testing

**Comment by nagilson** [comment]
I believe we had mentioned there was also an issue with the version completion, and that it should be tested in scenarios without a `nuget.config`.

**Comment by zivkan** [comment]
Was this tested with a private Azure DevOps feed?

By default, apps using NuGet.Protocol, `SourceRepository` will not look for credential providers. A static method needs to be called, and then `SourceRepository` will search for credential providers when it makes requests on feeds that need authentication.  Unfortunately, I can't remember the name of the API, so I can't quickly search the code to see if it's being called. But if testing showed that private Azure Artifacts feeds don't work, let me know and I'll look it up.

**Comment by baronfel** [comment]
It's worse than that @zivkan - the AzDo feeds don't actually support the searchautocompleteservice resource. The Registry apis will give you and AutoCompleteResource but then blow up with a protocolexception when you try to use that resource due to lack of the necessary endpoint. 

I would have expected a null instance if the Source didn't actually support autocompletion.

**Review by nagilson — APPROVED** [review]
Turns out the previous 'issues' were with the `nuget.config`... this is pretty awesome! Thank you for having a clean code style and writing useful comments. I could not find any bugs. I am very excited to see this feature ship 🚀

**Pull request merged**